In [ ]:
import numpy as np
import pandas as pd

In [ ]:
# load dataset

df = pd.read_csv('https://raw.githubusercontent.com/renatomaaliw3/public_files/refs/heads/master/Data%20Sets/NLP/moviereviews2.tsv', sep = '\t')
df

,label,review
0,pos,I loved this movie and will watch it again. Or...
1,pos,"A warm, touching movie that has a fantasy-like..."
2,pos,I was not expecting the powerful filmmaking ex...
3,neg,"This so-called ""documentary"" tries to tell tha..."
4,pos,This show has been my escape from reality for ...
...,...,...
5995,pos,"Of the three remakes of this plot, I like them..."
5996,neg,Poor Whoopi Goldberg. Imagine her at a friend'...
5997,neg,"Honestly before I watched this movie, I had he..."
5998,pos,This movie is essentially shot on a hand held ...


In [ ]:
# Check for missing values

df.isnull().sum()

,0
label,0
review,20


In [ ]:
# Remove NaN reviews

df.dropna(inplace = True)
len(df)

5980

In [ ]:
df.isnull().sum()

,0
label,0
review,0


In [ ]:
# remove blank reviews

blanks = []

for i, label, reviews in df.itertuples():

  if reviews.isspace():

    blanks.append(i)

blanks

[]

In [ ]:
# Remove blanks based from index

df.drop(blanks, inplace = True)
len(df)

5980

In [ ]:
df['label'].value_counts()

,count
label,
pos,2990
neg,2990


In [ ]:
# Train-Test Split

from sklearn.model_selection import train_test_split

X = df['review']
y = df['label']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 101)

In [ ]:
# Pipeline for Vectorization

from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import LinearSVC

text_clf = Pipeline([('tfidf', TfidfVectorizer()),('clf', LinearSVC())])
text_clf.fit(X_train, y_train)

Pipeline(steps=[('tfidf', TfidfVectorizer()), ('clf', LinearSVC())])

In [ ]:
# Predictions

predictions = text_clf.predict(X_test)
predictions

array(['pos', 'neg', 'pos', ..., 'neg', 'pos', 'neg'], dtype=object)

In [ ]:
# Metrics

from sklearn.metrics import confusion_matrix, classification_report, accuracy_score

print(confusion_matrix(y_test, predictions))

[[564  35]
 [ 36 561]]


In [ ]:
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

         neg       0.94      0.94      0.94       599
         pos       0.94      0.94      0.94       597

    accuracy                           0.94      1196
   macro avg       0.94      0.94      0.94      1196
weighted avg       0.94      0.94      0.94      1196



In [ ]:
print(accuracy_score(y_test, predictions))

0.9406354515050167


#01. What is the text classification accuracy of the model using LinearSVC?

# = 0.94

In [ ]:
# Pipeline for Vectorization
from sklearn.svm import SVC

text_clf = Pipeline([('tfidf', TfidfVectorizer()),('clf', SVC())])
text_clf.fit(X_train, y_train)

Pipeline(steps=[('tfidf', TfidfVectorizer()), ('clf', SVC())])

In [ ]:
predictions = text_clf.predict(X_test)

In [ ]:
print(confusion_matrix(y_test, predictions))

[[562  37]
 [ 44 553]]


In [ ]:
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

         neg       0.93      0.94      0.93       599
         pos       0.94      0.93      0.93       597

    accuracy                           0.93      1196
   macro avg       0.93      0.93      0.93      1196
weighted avg       0.93      0.93      0.93      1196



In [ ]:
print(accuracy_score(y_test, predictions))

0.9322742474916388


#02. What is the text accuracy of the model using SVC?

# = 0.93

In [ ]:
import nltk

In [ ]:
nltk.download('vader_lexicon')

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


True

In [ ]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer
sid = SentimentIntensityAnalyzer()

In [ ]:
df['scores'] = df['review'].apply(lambda review: sid.polarity_scores(review))
df

,label,review,scores
0,pos,I loved this movie and will watch it again. Or...,"{'neg': 0.062, 'neu': 0.695, 'pos': 0.243, 'co..."
1,pos,"A warm, touching movie that has a fantasy-like...","{'neg': 0.033, 'neu': 0.783, 'pos': 0.184, 'co..."
2,pos,I was not expecting the powerful filmmaking ex...,"{'neg': 0.097, 'neu': 0.795, 'pos': 0.108, 'co..."
3,neg,"This so-called ""documentary"" tries to tell tha...","{'neg': 0.116, 'neu': 0.832, 'pos': 0.052, 'co..."
4,pos,This show has been my escape from reality for ...,"{'neg': 0.028, 'neu': 0.769, 'pos': 0.203, 'co..."
...,...,...,...
5995,pos,"Of the three remakes of this plot, I like them...","{'neg': 0.028, 'neu': 0.825, 'pos': 0.148, 'co..."
5996,neg,Poor Whoopi Goldberg. Imagine her at a friend'...,"{'neg': 0.181, 'neu': 0.776, 'pos': 0.043, 'co..."
5997,neg,"Honestly before I watched this movie, I had he...","{'neg': 0.064, 'neu': 0.762, 'pos': 0.174, 'co..."
5998,pos,This movie is essentially shot on a hand held ...,"{'neg': 0.073, 'neu': 0.862, 'pos': 0.065, 'co..."


In [ ]:
df['compound'] = df['scores'].apply(lambda d: d['compound'])
df

,label,review,scores,compound
0,pos,I loved this movie and will watch it again. Or...,"{'neg': 0.062, 'neu': 0.695, 'pos': 0.243, 'co...",0.8720
1,pos,"A warm, touching movie that has a fantasy-like...","{'neg': 0.033, 'neu': 0.783, 'pos': 0.184, 'co...",0.9549
2,pos,I was not expecting the powerful filmmaking ex...,"{'neg': 0.097, 'neu': 0.795, 'pos': 0.108, 'co...",0.7201
3,neg,"This so-called ""documentary"" tries to tell tha...","{'neg': 0.116, 'neu': 0.832, 'pos': 0.052, 'co...",-0.9821
4,pos,This show has been my escape from reality for ...,"{'neg': 0.028, 'neu': 0.769, 'pos': 0.203, 'co...",0.9935
...,...,...,...,...
5995,pos,"Of the three remakes of this plot, I like them...","{'neg': 0.028, 'neu': 0.825, 'pos': 0.148, 'co...",0.9377
5996,neg,Poor Whoopi Goldberg. Imagine her at a friend'...,"{'neg': 0.181, 'neu': 0.776, 'pos': 0.043, 'co...",-0.9578
5997,neg,"Honestly before I watched this movie, I had he...","{'neg': 0.064, 'neu': 0.762, 'pos': 0.174, 'co...",0.9878
5998,pos,This movie is essentially shot on a hand held ...,"{'neg': 0.073, 'neu': 0.862, 'pos': 0.065, 'co...",-0.1010


In [ ]:
# Adding Neutral

def review_rating(scores):

  if scores == 0: # you can define a range -0.2 to 0.2

    return 'neu'

  elif scores > 0:

    return 'pos'

  else:

    return 'neg'

In [ ]:
df['comp_score'] = df['compound'].apply(review_rating)
df

,label,review,scores,compound,comp_score
0,pos,I loved this movie and will watch it again. Or...,"{'neg': 0.062, 'neu': 0.695, 'pos': 0.243, 'co...",0.8720,pos
1,pos,"A warm, touching movie that has a fantasy-like...","{'neg': 0.033, 'neu': 0.783, 'pos': 0.184, 'co...",0.9549,pos
2,pos,I was not expecting the powerful filmmaking ex...,"{'neg': 0.097, 'neu': 0.795, 'pos': 0.108, 'co...",0.7201,pos
3,neg,"This so-called ""documentary"" tries to tell tha...","{'neg': 0.116, 'neu': 0.832, 'pos': 0.052, 'co...",-0.9821,neg
4,pos,This show has been my escape from reality for ...,"{'neg': 0.028, 'neu': 0.769, 'pos': 0.203, 'co...",0.9935,pos
...,...,...,...,...,...
5995,pos,"Of the three remakes of this plot, I like them...","{'neg': 0.028, 'neu': 0.825, 'pos': 0.148, 'co...",0.9377,pos
5996,neg,Poor Whoopi Goldberg. Imagine her at a friend'...,"{'neg': 0.181, 'neu': 0.776, 'pos': 0.043, 'co...",-0.9578,neg
5997,neg,"Honestly before I watched this movie, I had he...","{'neg': 0.064, 'neu': 0.762, 'pos': 0.174, 'co...",0.9878,pos
5998,pos,This movie is essentially shot on a hand held ...,"{'neg': 0.073, 'neu': 0.862, 'pos': 0.065, 'co...",-0.1010,neg


In [ ]:
print(accuracy_score(df['label'], df['comp_score']))

0.7316053511705686


In [ ]:
print(classification_report(df['label'], df['comp_score']))

              precision    recall  f1-score   support

         neg       0.83      0.58      0.68      2990
         neu       0.00      0.00      0.00         0
         pos       0.68      0.88      0.77      2990

    accuracy                           0.73      5980
   macro avg       0.50      0.49      0.48      5980
weighted avg       0.76      0.73      0.73      5980



/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


#03. Using VADER, what is the overall accuracy of the model in terms of sentiment analysis?

# = 0.73